File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>>res;

        for(int i=0;i<nums.size();i++){
            if(nums[i]>0)break;
            if(i>0 && nums[i] == nums[i-1])continue;

            int l = i+1;
            int r = nums.size()-1;

            while(l<r){
                int csum = nums[i] + nums[l] + nums[r];
                if(csum > 0){
                    r--;
                }else if(csum < 0){
                    l++;
                }else{
                    res.push_back({nums[i],nums[l],nums[r]});
                    r--;
                    l++;
                    while(l<r && nums[l] == nums[l-1]){
                        l++;
                    }
                }
            }

        }
        return res;

    }
};

File: Data Structures & Algorithms/subsets-ii/submission-0.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        curr= []
        res = []


        def subset(index,nums,curr,res):

            if index >= len(nums):
                res.append(list(curr))
                return

            curr.append(nums[index])
            subset(index+1,nums,curr,res)
            curr.pop()

            next_index = index+1
            while next_index < len(nums) and nums[next_index] == nums[index]:
                next_index+=1

            subset(next_index,nums,curr,res)

        subset(0,nums,curr,res)
        return res